### PR TITLE
add support for signed media (bsc#1139561)

### PR DESCRIPTION
**Pull request by wfeldt**
### Problem

Original issue: https://bugzilla.suse.com/show_bug.cgi?id=1139561

checkmedia (correctly) reports wrong checksum of SUSE media. This was caused by our build tools that sign some files **after** Kiwi has added the checksums. That has been fixed (the checksum is adjusted after signing).

### New feature

Along with fixing the above issue our build tools now also sign the checksums stored on the medium. This allows checkmedia to verify SUSE media integrity by checking this signature.

### See

https://github.com/openSUSE/mksusecd/pull/42

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,28 +22,57 @@ the result against the stored digest.
...
+For this, a tag 'signature' is added pointing to a 2 kiB block to be used
+for the gpg signature of the 512 bytes application_data block. The tag is
+automatically added during digest calculation (`tagmedia --digest`). But you need to
+add the actual signature later.
+
+To create signed media, use `tagmedia --export-tags foo` to export the tag
+block to file `foo`. Then create a detached signature with gpg (`foo.asc`)
+and add the signature to the medium with `tagmedia --import-signature foo.asc`.
+
+To verify, ensure the public part of the gpg key is in `/usr/lib/rpm/gnupg/keys`.
```
Comment (lslezak) [inline README.md]: That means if I want to verify the signature of a 3rd party medium I have to import the unknown GPG key to my system?

:thinking: Hm, I do not like that much.... Would it be possible to use a GPG key without importing it first?

File: mediacheck.c [hunk 2]
```diff
@@ -769,3 +799,160 @@ void process_chunk(mediacheck_digest_t *digest, chunk_region_t *region, unsigned
...
+ * Verify signature.
+ *
+ * Call mediacheck_init() before doing this.
+ *
+ * The is function imports all keys from /usr/lib/rpm/gnupg/keys into a
+ * temporary key ring and then runs gpg to verify the signature.
+ */
+void verify_signature(mediacheck_t *media)
+{
+  char tmp_dir[] = "/tmp/mediacheck.XXXXXX";
```
Comment (lslezak) [inline mediacheck.c:889]: NP: Maybe we should prefer `$TMPDIR` and use `/tmp` as a fallback when not set, like `mktemp`... ?

File: mediacheck.c [hunk 3]
```diff
@@ -769,3 +799,160 @@ void process_chunk(mediacheck_digest_t *digest, chunk_region_t *region, unsigned
...
+
+  if((f = fopen(buf, "w"))) {
+    fprintf(f, "%s", media->signature.data);
+    fclose(f);
+  }
+
+  free(buf);
+
+  asprintf(&buf,
+    "/usr/bin/gpg --batch --homedir %s --no-default-keyring --ignore-time-conflict --ignore-valid-from "
```
Comment (lslezak) [inline mediacheck.c:917]: Do we need to escape the `%s` values or is it safe to not contain any special chars?

File: tagmedia [hunk 4]
```diff
@@ -59,6 +59,13 @@ use constant {
   MBR_LENGTH => 0x200,
 };
 
+# Some SUSE specific constants.
+#
+use constant {
+  # signature block starts with this string
+  SIGNATURE_MAGIC => "7984fc91-a43f-4e45-bf27-6d3aa08b24cf"
```
Comment (lslezak) [inline tagmedia:70]: Q: Where this comes from? Just in case we need to update that...

**Review by lslezak — COMMENTED** [review]
Just some question after a quick scan...

File: README.md (hunk 1 above)
Comment (wfeldt) [reply README.md]: Yes, I was unsure here.

How about adding an option like `--key-file FILE` to point to a public key. If set, this key is used for verification.

File: mediacheck.c (hunk 2 above)
Comment (wfeldt) [reply mediacheck.c:889]: I wouldn't want to rely on an environment var. `/tmp` is always a suitable dir.

File: mediacheck.c (hunk 3 above)
Comment (wfeldt) [reply mediacheck.c:917]: No need. `tmp_dir` above is fully under our control.

File: tagmedia (hunk 4 above)
Comment (wfeldt) [reply tagmedia:70]: added links to code

File: README.md (hunk 1 above)
Comment (wfeldt) [reply README.md]: implemented something

Comment (lslezak) [reply README.md]: Yes, some extra option would be useful.

File: mediacheck.c (hunk 2 above)
Comment (lslezak) [reply mediacheck.c:889]: OK

**Comment by shundhammer** [comment]
LGTM

**Pull request merged**